Remove locks from COM events delegate management.

AaronRobinsonMSFT opened the pull request:
This removes locks and instead assumes the collection is immutable.

Fixes #73754

AaronRobinsonMSFT committed:
Remove locks from COM events delegate management.

This removes locks and instead assumes the collection is immutable.

AaronRobinsonMSFT commented:
/cc @dotnet/interop-contrib

stephentoub commented on src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs:
```diff
@@ -99,7 +100,7 @@ private void PreProcessSignature()
         /// we need to complement this design by using an explicit linked list data structure.
         /// </summary>
-        private readonly List<DelegateWrapper> _delegateWrappers = new List<DelegateWrapper>();
+        private List<DelegateWrapper> _delegateWrappers = new List<DelegateWrapper>();
```
Does `List<T>` by you much at this point?  Now that you're creating a new instance each time, adding is allocate+copy+store and remove is allocate+copy+copy.  You might slim it down slightly by just using an array directly.

AaronRobinsonMSFT replied:
That is true, but then we do need to handle the removal case and that can be far more cumbersome. The add is simple, but removal requires me to copy, null and then shift > delete index down by one. I can do that but I think it is going to require more code unless there is as simple - remove and shift for arrays?

stephentoub replied:
```C#
var newArr = new DelegateWrapper[oldArr.Length - 1];
oldArr.AsSpan(0, i).CopyTo(newArr);
oldArr.AsSpan(i + 1).CopyTo(newArr.AsSpan(i));
```
? Regardless, it's up to you.  I don't know how much the extra allocation for lists matter here, especially since we're creating a new array now on every mutation.

davidfowl replied:
ImmutableArray?

jkoritzinsky replied:
I don’t believe that’s available at the later this code compiled into (it’s compiled into CoreLib amongst other places)

AaronRobinsonMSFT replied:
> ```csharp
> var newArr = new DelegateWrapper[oldArr.Length - 1];
> oldArr.AsSpan(0, i).CopyTo(newArr);
> oldArr.AsSpan(i + 1).CopyTo(newArr.AsSpan(i));
> ```

Ah. I regularly forget going to `Span` for these sorts of things.

stephentoub replied:
(It could also just use Array.Copy instead of AsSpan(...).Copy.)

stephentoub commented on src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs:
```diff
@@ -156,19 +157,21 @@ public bool Empty
-                {
-                    return _delegateWrappers.Count == 0;
-                }
+                var wrappers = _delegateWrappers;
```
```suggestion
                List<DelegateWrapper> wrappers = _delegateWrappers;
```

stephentoub commented on src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs:
```diff
@@ -246,27 +260,38 @@ public void RemoveDelegates(Func<Delegate, bool> condition)
             object? result = null;
 
-            lock (_delegateWrappers)
+            var wrappers = _delegateWrappers;
```
```suggestion
            List<DelegateWrapper> wrappers = _delegateWrappers;
```

AaronRobinsonMSFT committed:
Remove uses of var.

AaronRobinsonMSFT committed:
Use array instead of List<T>

AaronRobinsonMSFT committed:
Revert "Use array instead of List<T>"

This reverts commit 5ba348f0a70ced13d8abc3d5fe38be398ad98154.

AaronRobinsonMSFT commented:
@stephentoub The CI is green. Any other thoughts?

stephentoub commented on src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs:
```diff
@@ -177,21 +180,28 @@ public void AddDelegate(Delegate d, bool wrapArgs = false)
+                    DelegateWrapper wrapperMaybe = wrappers[i];
                     if (wrapperMaybe.Delegate.GetType() == d.GetType() && wrapperMaybe.WrapArgs == wrapArgs)
                     {
                         removeIdx = i;
```
Mainly I'm curious: are enough COM events subscribed to by multiple delegates in order to warrant this optimization of using Delegate.Combine rather than just having each delegate be its own entry in the list?  It looks safe from a cursory inspection, but it's a lot easier to prove correctness when the list being used is entirely immutable, whereas here the published list instance is having its contents modified.

jkotas replied:
I think there is a bug when `Remove` is called on two threads in parallel:

First thread starts creates a new array with the delegate removed
Second thread removes the delegate in-place without creating a new array
First thread swaps in its array, reverting the affect of the second remove.

stephentoub replied:
> I think there is a bug when Remove is called on two threads in parallel:

Yeah, this is an example of the kind of issue that comes from trying to get sneaky (I hadn't noticed this particular case, but I believe you're right).

I think it'd be better to just do the simple thing and actually have the list be immutable; then the only concurrency management need be the simple act of publishing safely.

Or if our goal is to stick with what netfx had, just revert the locking entirely and go back to exactly what netfx had :)

stephentoub commented on src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs:
```diff
@@ -246,27 +260,38 @@ public void RemoveDelegates(Func<Delegate, bool> condition)
                 {
-                    _delegateWrappers.RemoveAt(idx);
+                    newWrappers.RemoveAt(idx);
                 }
```
If you simplified the storage such that you didn't bother with Delegate.Combine, then it seems like tthis loop could be simplified as well to something along the lines of:
```C#
List<DelegateWrapper> oldWrappers, newWrappers;
do
{
    oldWrappers = _delegateWrappers;
    newWrappers  = new List<DelegateWrapper>(oldWrappers);
    if (newList.RemoveAll(w => condition(w.Delegate)) == 0)
    {
        return;
    }
}
while (!PublishNewWrappers(newWrappers, oldWrappers));
```

stephentoub commented on src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs:
```diff
@@ -99,7 +100,7 @@ private void PreProcessSignature()
         /// we need to complement this design by using an explicit linked list data structure.
         /// </summary>
-        private readonly List<DelegateWrapper> _delegateWrappers = new List<DelegateWrapper>();
+        private List<DelegateWrapper> _delegateWrappers = new List<DelegateWrapper>();
```
It seems like optimization of this code / reduction in allocation isn't needed.  But if we decide to in the future, there are various things that could be done.  For example, this empty list could be a singleton, or null could be special-cased to mean empty.  And if one handler is common, rather than being a list this could be an object, which could then store either a DelegateWrapper directly or a list of them.

AaronRobinsonMSFT replied:
Switched back to arrays and am now using `Array.Empty<T>` for this. It isn't non-allocating, but it is now a single allocation for this code path for process lifetime.

jkotas commented:
> Use array instead of List<T>
> Revert "Use array instead of List<T>" 

Was there an actual problem or you did not like the result? (It looks better to me with arrays instead of List.)

AaronRobinsonMSFT commented:
> > Use array instead of List
> > Revert "Use array instead of List"
> 
> Was there an actual problem or you did not like the result? (It looks better to me with arrays instead of List.)

It was an option. However, I needed to create a `RemoveAll` local with annoying logic so deferred to just using `List<T>`. Either works and appear correct, other than the "Remove" logic you pointed out above. Let me see about simplifying it further.

AaronRobinsonMSFT committed:
Revert "Revert "Use array instead of List<T>""

This reverts commit aa3dddd122920633e2c5e406ae92dc223c4a5f25.

AaronRobinsonMSFT committed:
Remove usage of `Delegate.Combine`. Upon deeper inspection there
doesn't seem to be any value to using that mechanism. An added
bonus is that it simplifies the logic.

AaronRobinsonMSFT commented:
@jkotas @stephentoub I switched back to using arrays, let me know if you prefer `List<T>`. I also removed the `Delegate.Combine` logic. My biggest concern with changing that related to threading since this is a COM code path. Looking through it, I no longer have any concerns so it was removed.

stephentoub commented on src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs:
```diff
@@ -206,50 +203,64 @@ public void RemoveDelegate(Delegate d, bool wrapArgs = false)
+                        toRemove.Add(i);
                     }
                 }
 
-                foreach (int idx in toRemove)
+                if (toRemove.Count == 0)
                 {
-                    _delegateWrappers.RemoveAt(idx);
+                    // Nothing to remove
+                    return;
                 }
+
+                newWrappers = RemoveAll(wrappers, CollectionsMarshal.AsSpan(toRemove));
+            } while (!PublishNewWrappers(newWrappers, wrappers));
+
+            // The list of indices is assumed to be sorted in ascending order
+            static DelegateWrapper[] RemoveAll(DelegateWrapper[] oldCol, ReadOnlySpan<int> toRemove)
+            {
+                // Allocate new collection
+                var newCol = new DelegateWrapper[oldCol.Length - toRemove.Length];
+                if (newCol.Length == 0)
+                {
+                    return newCol;
+                }
+
+                // Iterate over collection, skipping elements that should be removed.
+                int ri = 0;
+                for (int oi = 0, ni = 0; oi < oldCol.Length; oi++)
+                {
+                    if (ri < toRemove.Length && oi == toRemove[ri])
+                    {
+                        ri++;
+                        continue;
+                    }
+
+                    newCol[ni] = oldCol[oi];
+                    ni++;
+                }
+
+                return newCol;
```
Seems like this whole thing could just be:
```C#
DelegateWrapper[] oldWrappers, newWrappers;
do
{
    oldWrappers = _delegateWrappers;
    var tmp = new List<DelegateWrapper>(oldWrappers);
    tmp.RemoveAll(w => condition(w.Delegate));
    newWrappers = tmp.ToArray();
}
while (!PublishNewWrappers(newWrappers, oldWrappers));
```
I'm just not clear on what overheads we're trying to avoid and which we're not.

AaronRobinsonMSFT committed:
Use some sugar

AaronRobinsonMSFT committed:
Simplify the RemoveAll scenario.

stephentoub commented on src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs:
```diff
@@ -206,67 +201,44 @@ public void RemoveDelegate(Delegate d, bool wrapArgs = false)
             object? result = null;
 
-            lock (_delegateWrappers)
+            DelegateWrapper[] wrappers = _delegateWrappers;
+            foreach (DelegateWrapper wrapper in wrappers)
```
Nit: there's no need for the local here, it could just be:
```C#
foreach (DelegateWrapper wrapper in _delegateWrappers)
```
as was there previously, but if you prefer this style, it doesn't hurt

stephentoub commented on src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs:
```diff
@@ -156,46 +157,40 @@ public bool Empty
-                    return _delegateWrappers.Count == 0;
-                }
+                DelegateWrapper[] wrappers = _delegateWrappers;
+                return wrappers.Length == 0;
```
Nit: Could just be `get => _delegateWrappers.Length == 0;`, but if you prefer having the local, ok

stephentoub commented on src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs:
```diff
@@ -156,46 +157,40 @@ public bool Empty
         public void AddDelegate(Delegate d, bool wrapArgs = false)
         {
-            lock (_delegateWrappers)
+            DelegateWrapper[] wrappers;
+            DelegateWrapper[] newWrappers;
```
Nit: could be `DelegateWrapper[] wrappers, newWrappers;`

AaronRobinsonMSFT committed:
Style nits.

AaronRobinsonMSFT commented:
/backport to release/7.0

AaronRobinsonMSFT commented:
/backport to release/6.0
